llm/llama.go

```diff
@@ -218,366 +139,480 @@ func (ft llamaFileType) String() string {
+		params = append(params, "--mlock")
+	}
+	if !opts.UseMMap {
+		params = append(params, "--no-mmap")
+	}
+	if opts.UseNUMA {
+		params = append(params, "--numa")
 	}
 
-	llm.cursor = i
+	// try to start llama.cpp with gpu acceleration, if it fails, fallback to CPU
+	cmd, err := llamaCmd(llamaGPU, params)
+	if err != nil {
+		return nil, fmt.Errorf("llama.cpp gpu command setup: %w", err)
+	}
+	err = cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("error starting the external server: %w", err)
+	}
 
-	log.Printf("prompt: num_past=%d cached=%v eval=%v", i, len(llm.embd[:i]), len(llm.embd[i:]))
-	return cTokens
-}
+	// monitor the command in a goroutine, if it fails log the error and start the CPU version
+	go func() {
+		err := cmd.Wait()
+		if err != nil {
+			if err.Error() == "signal: killed" {
+				// this is expected when the server is closed due to loading a new model
+				return
+			}
+			// TODO: what is the specific error when the GPU is not supported?
+			log.Printf("could not start llama.cpp with gpu acceleration: %v\n", err)
+			// fallback to the CPU runner
+			cmd, err = llamaCmd(llamaCPU, params)
+			if err != nil {
+				log.Fatalf("error setting up the llama.cpp CPU runner: %v", err)
+			}
+			if err := cmd.Start(); err != nil {
+				log.Fatalf("error starting the llama.cpp CPU server: %v", err)
+			}
+		}
+	}()
 
-func (llm *llama) Encode(prompt string) []int {
-	cPrompt := C.CString(prompt)
-	defer C.free(unsafe.Pointer(cPrompt))
+	// wait for llama.cpp to come up
+	deadline := time.Now().Add(10 * time.Second)
 
-	cTokens := make([]C.llama_token, len(prompt)+1)
-	if n := C.llama_tokenize(llm.ctx, cPrompt, unsafe.SliceData(cTokens), C.int(len(cTokens)), true); n > 0 {
-		tokens := make([]int, n)
-		for i := range cTokens[:n] {
-			tokens[i] = int(cTokens[i])
+	for time.Now().Before(deadline) {
+		if !isPortAvailable(port) {
+			return &llama{Options: opts, Running: Running{Port: port, Cmd: cmd}}, nil
 		}
-
-		return tokens
+		time.Sleep(time.Millisecond)
 	}
 
-	return nil
+	return nil, fmt.Errorf("timed out waiting for llama.cpp at port %d", port)
 }
 
-func (llm *llama) Decode(tokens ...int) string {
-	var sb strings.Builder
-	for _, token := range tokens {
-		sb.WriteString(C.GoString(C.llama_token_to_str(llm.ctx, C.llama_token(token))))
-	}
+func (llm *llama) Close() {
+	llm.Running.Cmd.Process.Kill()
+}
 
-	return sb.String()
+func (llm *llama) SetOptions(opts api.Options) {
+	llm.Options = opts
 }
 
-func (llm *llama) next() (C.llama_token, error) {
-	llm.mu.Lock()
-	defer llm.mu.Unlock()
+type Prediction struct {
+	Content string `json:"content"`
+	Stop    bool   `json:"stop"`
+}
 
-	if len(llm.embd) >= llm.NumCtx {
-		numLeft := (llm.NumCtx - llm.NumKeep) / 2
-		truncated := llm.embd[:llm.NumKeep]
-		truncated = append(truncated, llm.embd[len(llm.embd)-numLeft:]...)
+type GenerationSettings struct {
+	FrequencyPenalty float64       `json:"frequency_penalty"`
+	Grammar          string        `json:"grammar"`
```

Comment:
This should be omitted for now

Reply:
I'm going to leave this one here so that the llama.cpp request has all its fields documented (same as the responses). We dont expose it yet.

Reply:
I would like to only have the options ollama supports rather than the ones llama.cpp supports. Otherwise it may cause confusion

Reply:
removed